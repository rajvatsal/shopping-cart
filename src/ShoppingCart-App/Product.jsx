const Product = ({ product, onImageLoad }) => {
  return (
    <div key={product.id}>
      <div title="name">{product.title}</div>
      <div title="description">{product.description}</div>
      <div title="category">{product.category}</div>
      <img src={product.image} alt="product image" onLoad={onImageLoad} />
      <div>{product.rating.rate}</div>
    </div>
  )
}

export default Product
