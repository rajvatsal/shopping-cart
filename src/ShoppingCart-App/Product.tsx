import classes from './Product.module.scss'
import { Link, useOutletContext } from 'react-router'
import { AppContext } from './App'
import * as api from '../ShoppingCart-Core/api'

const Product = ({ product }: { product: api.Product }) => {
  const { onImageLoad } = useOutletContext<AppContext>()

  return (
    <div key={product.id} className={classes.product}>
      <Link aria-label="product page" to={`product/${product.id}`}>
        <div className={classes.product__imgContainer}>
          <img src={product.image} alt="product image" onLoad={onImageLoad} />
        </div>
        <div title="category">{product.category}</div>
        <div title="name">{product.title}</div>
        <div className={classes.product__priceContainer}>
          <span title="price">${product.price}</span>
          <span title="rating">{product.rating.rate}</span>
        </div>
      </Link>
    </div>
  )
}

export default Product
