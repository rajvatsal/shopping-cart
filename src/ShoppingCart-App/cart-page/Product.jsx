import Counter from '../Counter.jsx'
import { Solid, Outline } from '../Icons/Bin.jsx'
import {
  product,
  product__toggleBtn,
  product__btnContainer,
  binOutline,
  binSolid,
} from './Product.module.scss'
import { useOutletContext } from 'react-router'
import { object } from 'prop-types'

const Product = ({ details }) => {
  const { updateProductCount } = useOutletContext()
  return (
    <div data-testid="product-cart-page" className={product}>
      <div>
        <h2 title="name">{details.title}</h2>
        <p title="price">${details.price}</p>
        <div className={product__btnContainer}>
          <Counter
            isInCart={true}
            id={details.id}
            updateValue={updateProductCount}
            value={details.count}
          />
          <button
            aria-label="remove"
            className={product__toggleBtn}
            onClick={() => {
              updateProductCount(details.id, 0)
            }}
          >
            <Solid className={binSolid} />
            <Outline className={binOutline} />
          </button>
        </div>
      </div>
      <img src={details.image} alt="product image" />
    </div>
  )
}

Product.propTypes = {
  details: object,
}

export default Product
