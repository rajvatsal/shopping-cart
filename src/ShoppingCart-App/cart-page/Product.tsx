import {
  pdt,
  pdt__toggleBtn,
  pdt__btnContainer,
  binOutline,
  binSolid,
} from './Product.module.scss'
import { useOutletContext } from 'react-router'
import { Solid, Outline } from '../Icons/Bin.tsx'
import { AppContext } from '../App.tsx'
import * as api from '../../ShoppingCart-Core/api.ts'
import Counter from '../Counter.tsx'

interface ProductCmp {
  product: api.Product
  count: number
}

const Product = ({ product, count }: ProductCmp) => {
  const { updateProductCount } = useOutletContext<AppContext>()
  return (
    <div data-testid="product-cart-page" className={pdt}>
      <div>
        <h2 title="name">{product.title}</h2>
        <p title="price">${product.price}</p>
        <div className={pdt__btnContainer}>
          <Counter
            ref={undefined}
            isInCart={true}
            id={product.id}
            value={count}
            updateValue={updateProductCount}
          />
          <button
            aria-label="remove"
            className={pdt__toggleBtn}
            onClick={() => {
              updateProductCount(product.id, 0, false)
            }}
          >
            <Solid className={binSolid} />
            <Outline className={binOutline} />
          </button>
        </div>
      </div>
      <img src={product.image} alt="product image" />
    </div>
  )
}

export default Product
