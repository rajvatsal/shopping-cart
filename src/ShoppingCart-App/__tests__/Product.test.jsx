import Product from '../Product.jsx'
import { products } from './mock-products.js'
import { render, fireEvent } from '@testing-library/react'

describe('Product Component', () => {
  it('DOM Structure', () => {
    const renderer = render(<Product product={products[0]} />)
    expect(renderer.container.firstChild).toMatchSnapshot()
  })

  it('Image Load Event', () => {
    const imageLoad = vi.fn()
    const { getByAltText } = render(
      <Product product={products[0]} onImageLoad={imageLoad} />
    )

    expect(imageLoad).toBeCalledTimes(0)

    fireEvent.load(getByAltText('product image'))

    expect(imageLoad).toBeCalledTimes(1)
  })
})
