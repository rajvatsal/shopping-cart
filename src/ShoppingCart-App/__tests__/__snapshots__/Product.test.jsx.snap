// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Product Component > DOM Structure 1`] = `
<div>
  <div
    title="name"
  >
    skateboard
  </div>
  <div
    title="description"
  >
    cool skateboard
  </div>
  <div
    title="category"
  >
    electronics
  </div>
  <img
    alt="product image"
    src="image-url"
  />
  <div />
</div>
`;
